### PR TITLE
Top addresses

### PR DESCRIPTION
**Pull request by flicker-harmony**
## Issue

Updated explorer service to collect top addresses and API /top for it.

## Test

Screenshot 2020-03-07 at 00 19 51

### Unit Test Coverage

Before:

```
<!-- copy/paste 'go test -cover' result in the directory you made change -->
```

After:

```
<!-- copy/paste 'go test -cover' result in the directory you made change -->
```

## Operational Checklist

1. **Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol?**. (If no, skip to question 8.)

   NO

2. **Describe the migration plan.**. For each flag epoch, describe what changes take place at the flag epoch, the anticipated interactions between upgraded/non-upgraded nodes, and any special operational considerations for the migration.

3. **Describe how the plan was tested.**

4. **How much minimum baking period after the last flag epoch should we allow on Pangaea before promotion onto mainnet?**

5. **What are the planned flag epoch numbers and their ETAs on Pangaea?**

6. **What are the planned flag epoch numbers and their ETAs on mainnet?**

    Note that this must be enough to cover baking period on Pangaea.

7. **What should node operators know about this planned change?**

8. **Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol?** (If no, continue to question 11.)

    NO

9. **Does the existing `node.sh` continue to work with this change?**

10. **What should node operators know about this change?**

11. **Does this PR introduce significant changes to the operational requirements of the node software, such as >20% increase in CPU, memory, and/or disk usage?**

    NO

### CODE REVIEW

File: api/service/explorer/service.go [hunk 1]
```diff
@@ -127,6 +150,108 @@ func (s *Service) GetAddresses(w http.ResponseWriter, r *http.Request) {
...
+		if err != nil {
+			utils.Logger().Error().Err(err).Msg("error reading epoch")
+			continue
+		}
+		epoch := big.NewInt(0)
+		if err = rlp.DecodeBytes(bytes, &epoch); err != nil {
+			utils.Logger().Error().Err(err).Msg("cannot decode epoch")
+			continue
+		}
+		for curEpoch := prevEpoch + 1; curEpoch < epoch.Int64(); curEpoch++ {
```
Comment (rlan35) [inline api/service/explorer/service.go]: don't need to loop over shard state validator addresses. Just loop through all addresses in the trie.

Comment (flicker-harmony) [reply api/service/explorer/service.go]: @rlan35 what trie? Which method to use?

**Comment by flicker-harmony** [comment]
Screenshot 2020-03-09 at 00 52 22

**Comment by rlan35** [comment]
Jenkins, test this please.

File: api/service/explorer/service.go [hunk 2]
```diff
@@ -127,6 +143,66 @@ func (s *Service) GetAddresses(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+// UpdateTopAddresses updates 20 top addresses by balance from storage each 10 minutes.
+func (s *Service) UpdateTopAddresses() {
+	ticker := time.NewTicker(10 * time.Minute)
```
Comment (fxfactorial) [inline api/service/explorer/service.go]: you never close the ticker and always leak the resource.

File: api/service/explorer/service.go [hunk 3]
```diff
@@ -127,6 +143,66 @@ func (s *Service) GetAddresses(w http.ResponseWriter, r *http.Request) {
...
+	utils.Logger().Info().Msg("start updating top addresses")
+	for range ticker.C {
+		db, err := s.State()
+		if err != nil {
+			utils.Logger().Error().Err(err).Msg("fetch state error")
+		}
+		it := trie.NewIterator(db.GetTrie().NodeIterator(nil))
+		topAddresses := make([]AddrBalance, 0)
```
Comment (fxfactorial) [inline api/service/explorer/service.go]: no purpose to use make when could use literal when size is 0

**Pull request merged**